src/main/kotlin/com/openlattice/batching/EntityBatchingService.kt

```diff
@@ -0,0 +1,167 @@
+    }
+
+    /**
+     * Adds up to [batchSize] id to the batch of type [batchType] in the [BATCHES] table.
+     *
+     * @param batchType Use group types of batch operations together. An example here might be indexing vs linking indexing
+     * @param idsSql A select sql fragment returns at least [ENTITY_SET_ID],[ID],[LINKING_ID], and [PARTITION] columns for batch entry that needs to be added.
+     * @param batchSize The maximum number of batch entries to retry to read from executing idsSql and insert.
+     * @param bind A bind function that takes a [PreparedStatement] and [Int] offset, performs the required binds for [idsSql] and returns the base after performing the binds.
+     */
+    fun enqueue(
+            batchType: String,
+            idsSql: String,
+            batchSize: Int = DEFAULT_BATCHING_SIZE,
+            bind: (PreparedStatement, Int) -> Int
+    ) {
+        hds.connection.use { connection ->
+            connection.prepareStatement(buildInsertBatch(idsSql)).use { ps ->
+                ps.setString(1, batchType)
+                val offset = bind(ps, 1)
+                ps.setInt(offset + 1, batchSize)
+            }
+        }
+    }
+
+    /**
+     * Counts the number of batch entries of a given batch type.
+     *
+     * @param batchType Use group types of batch operations together. An example here might be indexing vs linking indexing
+     */
+    fun count(batchType: String): Long = hds.connection.use { connection ->
+        connection.prepareStatement(COUNT_BATCH_TYPE).use { ps ->
+            ps.setString(1, batchType)
+            val rs = ps.executeQuery()
+            require(rs.next()) { "Something went wrong when counting for batch type $batchType" }
+            ResultSetAdapters.count(rs)
+        }
+    }
+
+    /**
+     * See [processBatch]
+     */
+    fun processBatch(
+            batchType: String,
+            batchSize: Int,
+            process: (List<EntityBatchEntry>) -> Unit
+    ) = processBatch<Unit>(batchType, batchSize, process)
+
+    /**
+     * @param batchType Use group types of batch operations together. An example here might be indexing vs linking indexing
+     * @param batchSize The maximum number of batch entries to process.
+     * @param process The function to applied to each batch.
+     */
+    fun <T> processBatch(
+            batchType: String,
+            batchSize: Int,
+            process: (List<EntityBatchEntry>) -> T
+    ): T = hds.connection.use { connection ->
+        connection.autoCommit = false
+        try {
+            var remaining = batchSize
+            val batch = mutableListOf<EntityBatchEntry>()
+
+            /**
+             * Since this is a skip locked query, it won't ever deadlock and will simply skipped locked rows to process
+             * unlocked rows.
+             */
+            val result = connection.prepareStatement(SELECT_BATCH).use { ps ->
```

Comment:
Any reason not to use a BasePostgresIterable here? Seems preferable to get all the built-in error handling there

Reply:
BasePostgresIterable acquires and closes connections, which is incompatible with the skip locks needed here.